# L5: Dependency parsing

Dependency parsing is the task of mapping a sentence to a formal representation of its syntactic structure in the form of a dependency tree, which consists of directed arcs between individual words (tokens). In the lab you will implement a dependency parser based on the arc-standard algorithm and the fixed-window model that you implemented in Lab&nbsp;L4.

## The data set

The data set for this lab is the same as for Lab&nbsp;L4: the English Web Treebank from the [Universal Dependencies Project](http://universaldependencies.org). The code below defines an iterable-style dataset for parser data in the [CoNLL-U format](https://universaldependencies.org/format.html) that the project uses to distribute its data.

In [ ]:
class Dataset():

    ROOT = ('<root>', '<root>', 0)  # Pseudo-root

    def __init__(self, filename):
        self.filename = filename

    def __iter__(self):
        with open(self.filename, 'rt', encoding='utf-8') as lines:
            tmp = [Dataset.ROOT]
            for line in lines:
                if not line.startswith('#'):  # Skip lines with comments
                    line = line.rstrip()
                    if line:
                        columns = line.split('\t')
                        if columns[0].isdigit():  # Skip range tokens
                            tmp.append((columns[1], columns[3], int(columns[6])))
                    else:
                        yield tmp
                        tmp = [Dataset.ROOT]

We load the training data and the development data:

In [ ]:
train_data = Dataset('en_ewt-ud-train-projectivized.conllu')
dev_data = Dataset('en_ewt-ud-dev.conllu')

Both data sets consist of **parsed sentences**. A parsed sentence is represented as a list of triples, where the first component of each triple (a string) represents a word, and the second component (also a string) represents the word’s part-of-speech tag. The third component (an integer) specifies the position of the word’s syntactic head, i.e., its parent in the dependency tree. Run the following code cell to see an example:

In [ ]:
example_sentence = list(train_data)[531]

example_sentence

[('<root>', '<root>', 0),
 ('I', 'PRON', 2),
 ('like', 'VERB', 0),
 ('yuor', 'PRON', 4),
 ('blog', 'NOUN', 2),
 ('.', 'PUNCT', 2)]

In this example the head of the pronoun *I* is the word at position&nbsp;2 – the verb *like*. The dependents of *like* are *I* (position&nbsp;1) and the noun *blog* (position&nbsp;4), as well as the final punctuation mark. Note that each sentence starts with the so-called **pseudo-root** (position&nbsp;0). This pseudo-root is a pseudo-word that is guaranteed to be the root of the dependency tree.

## Parser interface

Like the tagger in the previous lab, the parser that you will implement in this lab follows a simple interface:

In [ ]:
class Parser(object):

    def predict(self, words, tags):
        raise NotImplementedError

The single method of this interface has the following specification:

**predict** (*self*, *words*, *tags*)

> Returns the list of predicted heads (a list of integers) for a single sentence, specified in terms of its *words* (a list of strings) and their corresponding *tags* (also a list of strings).

One trivial implementation of this interface is a parser that attaches each (real) word to its preceding word:

In [ ]:
class TrivialParser(Parser):

    def predict(self, words, tags):
        return [0] + list(range(len(words)-1))

## Problem 1: Implement an evaluation function

Your first task is to implement a function that computes the **unlabelled attachment score (UAS)** of a parser on gold-standard data.

In [ ]:
def uas(parser, gold_data):
    correct = 0
    total = 0
    for data in gold_data:
      predict = parser.predict([t[0] for t in data], [t[1] for t in data])
      gold = [t[2] for t in data]
      correct += sum(a == b for a, b in zip(predict[1:len(predict)], gold[1:len(gold)]))
      total += len(predict)-1
    return correct/total

Your implementation should conform to the following specification:

**uas** (*parser*, *gold_data*)

> Computes the unlabelled attachment score of the specified *parser* on the gold-standard data *gold_data* (an iterable of tagged sentences) and returns it as a float. The unlabelled attachment score is the percentage of all tokens to which the parser assigns the correct head (as per the gold standard). The calculation excludes the pseudo-roots.

### 🤞 Test your code

Test your code by computing the unlabelled attachment score for the trivial parser that attaches every word to its preceding word. The expected score on the development set is 9.76%.

In [ ]:
parser = TrivialParser()

uas_score = uas(parser, list(dev_data))
print(uas_score)

0.09757843254204938


## Problem 2: Create the vocabularies

The next cell contains skeleton code for a function `make_vocabs` that constructs the two vocabularies of the parser: one for the words and one for the tags. You should be quite familiar with this task by now. You will be able to re-use your code from lab&nbsp;L4.

In [ ]:
PAD = '<pad>'
UNK = '<unk>'

def make_vocabs(gold_data):
    counter = []
    vocab_words = {}
    vocab_tags = {}
    vocab_words[PAD] = len(vocab_words)
    vocab_tags[PAD] = len(vocab_tags)
    vocab_words[UNK] = len(vocab_words)
    for data in gold_data:
      for word in data:
        if word[0] not in vocab_words:
          vocab_words[word[0]] = len(vocab_words)
        if word[1] not in vocab_tags:
          vocab_tags[word[1]] = len(vocab_tags)
    return vocab_words, vocab_tags

Complete the code according to the following specification:

**make_vocabs** (*gold_data*)

> Returns a pair of dictionaries mapping the unique words and tags in the gold-standard data *gold_data* (an iterable over parsed sentences) to contiguous ranges of integers starting at zero. The word dictionary contains the pseudowords `PAD` (index&nbsp;0) and `UNK` (index&nbsp;1); the tag dictionary contains `PAD` (index&nbsp;0).

### 🤞 Test your code

Test your implementation by computing the total number of unique words and part-of-speech tags in the training data (including the pseudowords and the part-of-speech tag for the pseudoroot). The expected values are 19,676&nbsp;words and 19&nbsp;tags.

In [ ]:
vocab_words, vocab_tags = make_vocabs(list(train_data))
print(len(vocab_words), len(vocab_tags))

19676 19


## Problem 3: Implement the arc-standard algorithm

The parser that you will implement in this lab consists of two parts: a static part that implements the logic of the arc-standard algorithm (presented in Lecture&nbsp;5.2), and a non-static part that contains the learning component – the fixed-window model that you implemented in Lab&nbsp;L4. In this problem you will implement the static part; the learning component is covered in Problem&nbsp;5.

Recall that, in the arc-standard algorithm, the next move (also called ‘transition’) of the parser is predicted based on features extracted from the current parser configuration, with references to the words and part-of-speech tags of the input sentence. On the Python side of things, the words and part-of-speech tags are represented as lists of strings, and a configuration is represented as a triple

$$
(i, \mathit{stack}, \mathit{heads})
$$

where $i$ is an integer specifying the position of the next word in the buffer, $\mathit{stack}$ is a list of integers specifying the positions of the words currently on the stack (with the topmost element last in the list), and $\mathit{heads}$ is a list of integers specifying the positions of the head words. If a word has not yet been assigned a head, its head value is&nbsp;0. To illustrate this representation, the initial configuration for the example sentence above is

and a possible final configuration is

**Note:** In Lecture&nbsp;5.2, both the buffer and the stack were presented as list of words. Here we only represent the *stack* as a list of words. To represent the *buffer*, we simply record the position of the next word that has not been processed yet (the integer $i$). This acknowledges the fact that the buffer (in contrast to the stack) can never grow, but will be processed from left to right.

The cell below contains a complete skeleton for the logic of the arc-standard algorithm:

In [ ]:
class ArcStandardParser(Parser):

    MOVES = tuple(range(3))

    SH, LA, RA = MOVES  # Parser moves are specified as integers.


    @staticmethod
    def initial_config(num_words):
        i = 0
        stack = []
        heads = [0 for _ in range(num_words)]
        return (i, stack, heads)


    @staticmethod
    def valid_moves(config):
        valid = [0,1,2]
        if len(config[1]) <= 1:
          SH = 1
          del valid[1]
          del valid[1]
        if config[0] >= len(config[2]):
          del valid[0]

        return valid

    @staticmethod
    def next_config(config, move):
        i = config[0]
        stack = config[1]
        heads = config[2]
        if move == 0:
          stack.append(i)
          i += 1
        elif move == 1:
          heads[stack[-2]] = stack[-1]
          del stack[-2]
        elif move == 2:
          heads[stack[-1]] = stack[-2]
          del stack[-1]
        return (i, stack, heads)

    @staticmethod
    def is_final_config(config):
        if len(parser.valid_moves(config)) == 0:
          return True
        return False

Your implementation should conform to the following specification:

**initial_config** (*num_words*)

> Returns the initial configuration for a sentence with the specified number of words (*num_words*).

**valid_moves** (*config*)

> Returns the list of valid moves for the specified configuration (*config*).

**next_config** (*config*, *move*)

> Applies the *move* in the specified configuration *config* and returns the new configuration. This must not modify the input configuration.

**is_final_config** (*config*)

> Tests whether *config* is a final configuration.

### 🤞 Test your code

To test your implementation, you can run the code below. The code in this cell creates the initial configuration for the example sentence, simulates a sequence of moves, and then tests that the resulting configuration is the expected final configuration.

In [ ]:
moves = [0, 0, 0, 1, 0, 0, 1, 2, 0, 2, 2]    # 0 = SH, 1 = LA, 2 = RA

parser = ArcStandardParser()
config = parser.initial_config(len(example_sentence))
for move in moves:
    assert move in parser.valid_moves(config)
    config = parser.next_config(config, move)
assert parser.is_final_config(config)
assert config == (6, [0], [0, 2, 0, 4, 2, 2])

print('Looks good!')

Looks good!


## Problem 4: Implement the oracle

The learning component of the parser is the next move classifier. To train this classifier, we need training examples of the form $(\mathbf{x}, m)$, where $\mathbf{x}$ is a feature vector extracted from a given parser configuration $c$, and $m$ is the corresponding gold-standard move. To obtain $m$, we need an **oracle**.

Recall that, in the context of transition-based dependency parsing, an oracle is a function that translates a gold-standard dependency tree (here represented as a list of head ids) into a sequence of moves such that, when the parser takes the moves starting from the initial configuration, then it recreates the original dependency tree. Here we ask you to implement the static oracle that was presented in Lecture&nbsp;5.2.

In [ ]:
def oracle_moves(gold_heads):

    parser = ArcStandardParser()
    config = parser.initial_config(len(gold_heads))
    moves = []

    for i in range(2*len(gold_heads)-1):
        valid_moves = parser.valid_moves(config)
        if len(valid_moves) == 1:
          yield config, valid_moves[0]
          config =  parser.next_config(config, valid_moves[0])
          moves.append(valid_moves[0])

        else:
          # choose LA
          counterGoldSecond = [1 for i in gold_heads if i ==  config[1][-2]]
          counterHeadsSecond = [1 for i in config[2] if i ==  config[1][-2]]

          # choose RA
          counterGoldTop = [1 for i in gold_heads if i ==  config[1][-1]]
          counterHeadsTop = [1 for i in config[2] if i ==  config[1][-1]]

          if (len(counterGoldSecond) == len(counterHeadsSecond)) and (gold_heads[config[1][-2]] == config[1][-1]):
            yield config, 1
            config =  parser.next_config(config, 1)
            moves.append(1)

          elif (len(counterGoldTop) == len(counterHeadsTop)) and (gold_heads[config[1][-1]] == config[1][-2]):
            yield config, 2
            config =  parser.next_config(config, 2)
            moves.append(2)

          else:
            yield config, 0
            config =  parser.next_config(config, 0)
            moves.append(0)


Your implementation should conform to the following specification:

**oracle_moves** (*gold_heads*)

> Translates a gold-standard head assignment for a single sentence (*gold_heads*) into the corresponding stream of oracle moves. More specifically, this yields pairs $(c, m)$ where $m$ is a move (an integer, as specified in the `ArcStandardParser` interface) and $c$ is the parser configuration in which $m$ was taken.

### 🤞 Test your code

Test your code by running the cell below. This uses your implementation of *oracle_moves* to extract the oracle move sequence from the example sentence and compares it to the gold-standard move sequence *gold_moves*.

In [ ]:
gold_heads = [h for w, t, h in example_sentence]
gold_moves = [0, 0, 0, 1, 0, 0, 1, 2, 0, 2, 2]

assert list(m for _, m in oracle_moves(gold_heads)) == gold_moves

## Problem 5: Fixed-window parser

Now it is time to put everything together. For the full implementation of the fixed-window parser, you will need the correspondents of the four parts of the fixed-window tagger from Lab&nbsp;L4: an implementation of the fixed-window model; a parser that uses the fixed-window model to make predictions; a function that generates the training examples for the parser; and the training loop.

### Problem 5.1: Implement the fixed-window model

The fixed-window model for the parser is the same as the fixed-window model for the tagger in Lab&nbsp;L4. You can simply copy your code from that lab.

In [ ]:
import torch.nn as nn
import torch
class FixedWindowModel(nn.Module):

    def __init__(self, embedding_specs, hidden_dim, output_dim):
        super().__init__()
        moduleList = nn.ModuleList([])
        self.embedding_layers = nn.ModuleList([])
        self.linear_input = 0
        n_words = 0
        n_tags = 0
        self.k = 0
        index = 1
        for specs in embedding_specs:
          if index == 0:
            embedding = nn.Embedding(specs[1], specs[2])
            embedding.weight.data.normal_(0.0,0.01)
            index += 1
          else:
              embedding = nn.Embedding(specs[1], specs[2])
              embedding.weight.data.normal_(0.0,0.01)
          self.embedding_layers.append(embedding)
          self.linear_input += specs[0]*specs[2]
          self.k += specs[0]
          if specs[1] > n_words:
            n_words = specs[1]
          if specs[1] != n_words:
            n_tags = specs[1]
        self.instances = [value[0] for value in embedding_specs]
        self.hidden_linear = nn.Linear(self.linear_input, hidden_dim, bias = True)
        self.relu = torch.relu
        self.linear = nn.Linear(hidden_dim, 3, bias = True)

    def forward(self, features):
        index = 0
        embedd = []
        for i in range(len(self.embedding_layers)):
          embedd.append(self.embedding_layers[i](features[:,index:index+self.instances[i]].long()))
          index += self.instances[i]
          if self.instances[i] > 1 :
            embedd[i] = embedd[i].view(len(features),1,-1)
        concat = torch.cat(tuple(embedd), dim=2)
        hidden = self.hidden_linear(concat)
        relu = self.relu(hidden)
        output = self.linear(relu)
        return output

### Problem 5.2: Implement the parser

The next step is to implement the parser itself. This parser will use the fixed-window model to predict the next move for a given configuration in the arc-standard algorithm, based on the features extracted from the current feature window.

#### Default feature model

For the parser, we ask you to implement a fixed-window model with the following features ($k=6$):

0. word form of the next word in the buffer
1. word form of the topmost word on the stack
2. word form of the second-topmost word on the stack
3. part-of-speech tag of the next word in the buffer
4. part-of-speech tag of the topmost word on the stack
5. part-of-speech tag of the second-topmost word on the stack

Whenever the value of a feature is undefined, you should use the special value `PAD`.

#### Hyperparameters

The following choices are reasonable defaults for the hyperparameters of the network architecture used by the parser:

* width of the word embedding: 50
* width of the tag embedding: 10
* size of the hidden layer: 180

In [ ]:
import math
class FixedWindowParser(ArcStandardParser):

    def __init__(self, vocab_words, vocab_tags, word_dim=50, tag_dim=10, hidden_dim=180):
        self.embedding_specs = [(3, len(vocab_words), word_dim), (3, len(vocab_tags), tag_dim)]
        self.model = FixedWindowModel(self.embedding_specs, len(vocab_tags), hidden_dim)
        self.vocab_words = vocab_words
        self.vocab_tags = vocab_tags

    def featurize(self, words, tags, config):
        output = torch.zeros(6)
        # Add next word and tag in buffer
        if config[0] < len(words):
          output[0] = words[config[0]]
          output[3] = tags[config[0]]
        else:
          output[0] = self.vocab_words['<pad>']
          output[3] = self.vocab_tags['<pad>']
        # Add top most and second top words and their tags from stack
        if len(config[1]) > 1:
          output[1] = words[config[1][-1]]
          output[4] = tags[config[1][-1]]
          output[2] = words[config[1][-2]]
          output[5] = tags[config[1][-2]]
        elif len(config[1]) > 0:
          output[1] = words[config[1][-1]]
          output[4] = tags[config[1][-1]]
          output[2] = self.vocab_words['<pad>']
          output[5] = self.vocab_tags['<pad>']
        else:
          output[1] = self.vocab_words['<pad>']
          output[4] = self.vocab_tags['<pad>']
          output[2] = self.vocab_words['<pad>']
          output[5] = self.vocab_tags['<pad>']
        return output

    def predict(self, words, tags):
        words_id = torch.zeros(len(words))
        tags_id = torch.zeros(len(tags))
        for i in range(len(words)):
          tags_id[i] = self.vocab_tags[tags[i]]
          if words[i] in self.vocab_words:
            words_id[i] = self.vocab_words[words[i]]
          else:
            words_id[i] = self.vocab_words['<unk>']
        parser = ArcStandardParser()
        config = parser.initial_config(len(words))
        while not parser.is_final_config(config):
          features = self.featurize(words_id, tags_id, config)
          features = features.unsqueeze(0)
          moves = self.model.forward(features)
          moves = moves.squeeze(0)
          moves = moves.squeeze(0)
          best_move = None
          highest_score = - math.inf
          valid_moves = parser.valid_moves(config)
          for i in range(len(moves)):
            if moves[i] > highest_score:
              if i in valid_moves:
                highest_score = moves[i]
                best_move = i
          config = parser.next_config(config, best_move)
        return config[2]

Complete the skeleton code by implementing the methods of this interface:

**__init__** (*self*, *vocab_words*, *vocab_tags*, *word_dim* = 50, *tag_dim* = 10, *hidden_dim* = 100)

> Creates a new fixed-window model of appropriate dimensions and sets up any other data structures that you consider relevant. The parameters *vocab_words* and *vocab_tags* are the word vocabulary and tag vocabulary. The parameters *word_dim* and *tag_dim* specify the embedding width for the word embeddings and tag embeddings.

**featurize** (*self*, *words*, *tags*, *config*)

> Extracts features from the specified parser state according to the feature model given above. The state is specified in terms of the words in the input sentence (*words*, a list of word ids), their part-of-speech tags (*tags*, a list of tag ids), and the parser configuration proper (*config*, as specified in Problem&nbsp;3).

**predict** (*self*, *words*, *tags*)

> Predicts the list of all heads for the input sentence. This simulates the arc-standard algorithm, calling the move classifier whenever it needs to take a decision. The input sentence is specified in terms of the list of its words (strings) and the list of its tags (strings). Both of these should include the pseudoroot.

#### 💡 Hint on the implementation

In the *predict* function, you must make sure to only execute valid moves. One simple way to do so is to let the fixed-window model predict scores for all moves, and to implement your own, customised argmax operation to find the *valid* move with the highest score.

### Problem 5.3: Generate the training examples

Your next task is to implement a function that generates the training examples for the parser. You will train as usual, using minibatch training.

In [ ]:
def training_examples(vocab_words, vocab_tags, gold_data, parser, batch_size=100):
    batch = torch.zeros((batch_size,6))
    moves = torch.zeros(batch_size)
    index = 0
    for data in gold_data:
      gold_heads = [t[2] for t in data]
     # print(gold_heads)
      words_id = torch.zeros(len(data))
      tags_id = torch.zeros(len(data))
      for i, word in enumerate(data):
        if word[0] in vocab_words:
          words_id[i] = vocab_words[word[0]]
        else:
          words_id[i] = vocab_words['<unk>']
        tags_id[i] = vocab_tags[word[1]]
      config = parser.initial_config(len(words_id))
      for config, move in oracle_moves(gold_heads):
        row = parser.featurize(words_id, tags_id, config)
        moves[index] = move
        batch[index] = row
        if index == (batch_size - 1):
          yield batch, moves
          index = -1
          batch = torch.zeros((batch_size,6))
          moves = torch.zeros(batch_size)
        index += 1

    if index < batch_size:
      batch = batch[0:index]
      moves = moves[0:index]
      yield batch, moves


Your code should comply with the following specification:

**training_examples** (*vocab_words*, *vocab_tags*, *gold_data*, *tagger*, *batch_size* = 100)

> Iterates through the given *gold_data* (an iterable of parsed sentences), encodes it into word ids and tag ids using the specified vocabularies *vocab_words* and *vocab_tags*, and then yields batches of training examples for gradient-based training. Each batch contains *batch_size* examples, except for the last batch, which may contain fewer examples. Each example in the batch is created by a call to the `featurize` function of the *parser*.

### Problem 5.4: Training loop

The last piece of the puzzle is the training loop. This should be straightforward by now. Complete the skeleton code in the cell below:

In [ ]:
import torch.optim as optim
import torch.nn.functional as F

def train_fixed_window(train_data, n_epochs=1, batch_size=100, lr=1e-2):
    # TODO: Replace the next line with your own code
     # Initialize the model and tagger
    parser = FixedWindowParser(vocab_words, vocab_tags)

    # Initialize the optimizer. Here we use Adam rather than plain SGD
    optimizer = optim.Adam(parser.model.parameters(), lr=lr)

    for epoch in range(n_epochs):
        parser.model.train()
        for batch, moves in training_examples(vocab_words, vocab_tags, train_data, parser):
            optimizer.zero_grad()
            output = parser.model.forward(batch)
            output = output.squeeze(1)
            moves = moves.long()
            loss = F.cross_entropy(output, moves)
            loss.backward()
            optimizer.step()
    return parser


Here is the specification of the training function:

**train_fixed_window** (*train_data*, *n_epochs* = 1, *batch_size* = 100, *lr* = 1e-2)

> Trains a fixed-window parser from a set of training data *train_data* (an iterable over parsed sentences) using minibatch gradient descent and returns it. The parameters *n_epochs* and *batch_size* specify the number of training epochs and the minibatch size, respectively. Training uses the cross-entropy loss function and the [Adam optimizer](https://pytorch.org/docs/stable/optim.html#torch.optim.Adam) with learning rate *lr*.

The next code cell trains a tagger and evaluates it on the development data:

In [ ]:
parser = train_fixed_window(train_data, n_epochs=1)
print('{:.4f}'.format(uas(parser, dev_data)))

0.7176


**⚠️ Your submitted notebook must contain output demonstrating at least 68% UAS on the development set.**

## Problem 6: Predicted part-of-speech tags (reflection)

The data that you have used in this lab so far contains gold-standard part-of-speech tags, which makes the evaluation of your parser somewhat misleading: In a practical system (including the baseline for the standard project), one does not have access to gold-standard tags; instead one has to first tag the sentences with an automatic part-of-speech tagger.

The lab directory contains the following alternative versions of the two data for this lab:

* `en_ewt-ud-train-projectivized-retagged.conllu`
* `en_ewt-ud-dev-retagged.conllu`

In each of them, the gold-standard part-of-speech tags have been replaced by part-of-speech tags automatically predicted by the tagger from Lab&nbsp;L4.

Run an experiment to assess the effect that using predicted part-of-speech tags instead of gold-standard tags has on the unlabelled attachment score of your parser. Document your exploration in a short reflection piece (ca. 150&nbsp;words). Respond to the following prompts:

* How did you set up your experiment? What results did you get?
* Based on what you know about machine learning, did you expect your results? How do you explain them?
* What did you learn? How, exactly, did you learn it? Why does this learning matter?

We changed to the new files and then ran the model on this data. When using gold-standard part-of-speech tags we got an accuracy of 71.76% but when using automatically predicted part-of-speech tags we got an accuracy of 66.37%. This result was expected since the automatically predicted part-of-sspeech tags are not 100% accurate meaning we will misslead the model sometimes. In this lab we learned about oracle and how dependency parsing works. We also got the chance to convert a tagger into a parser, this was pleasent since we managed to make the last lab work proparly it was quite straight-forward to implement a parser since we only had to change small things.

**🥳 Congratulations on finishing the last lab in this course! 🥳**